File: cmd/sample/api/messages.py
"""
# cmd: python -m samples.messages
# title: Message Content Types
# description: 文字、圖片、影片內容區塊範例
"""

import click
from .client import get_client


@click.group()
def messages():
    """Message content type examples"""
    pass


@messages.command()
# cmd: python -m samples.messages text
@click.command()
def text():
    """文字內容區塊 (text content block)"""
    client = get_client()
    response = client.messages.create(
        model="MiniMax-M3",
        max_tokens=1024,
        messages=[
            {
                "role": "user",
                "content": [
                    {"type": "text", "text": "What is the capital of France?"},
                    {"type": "text", "text": "What about Germany?"},
                ],
            }
        ],
    )
    click.echo(f"Response: {response.content[0].text}")


@messages.command()
# cmd: python -m samples.messages image_url
@click.command()
def image_url():
    """
    圖片內容區塊 (M3 only)
    - 支援格式: JPEG/PNG/GIF/WEBP
    - 最大10 MB (512 MB via Files API)
    """
    client = get_client()
    response = client.messages.create(
        model="MiniMax-M3",
        max_tokens=1024,
        messages=[
            {
                "role": "user",
                "content": [
                    {"type": "text", "text": "What is in this image?"},
                    {
                        "type": "image",
                        "source": {
                            "type": "url",
                            "media_type": "image/jpeg",
                            "url": "https://example.com/photo.jpg",
                        },
                    },
                ],
            }
        ],
    )
    click.echo(f"Response: {response.content[0].text}")


@messages.command()
# cmd: python -m samples.messages image_base64
@click.command()
def image_base64():
    """圖片 base64 編碼"""
    client = get_client()
    response = client.messages.create(
        model="MiniMax-M3",
        max_tokens=1024,
        messages=[
            {
                "role": "user",
                "content": [
                    {"type": "text", "text": "Describe this image."},
                    {
                        "type": "image",
                        "source": {
                            "type": "base64",
                            "media_type": "image/png",
                            "data": "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg==",
                        },
                    },
                ],
            }
        ],
    )
    click.echo(f"Response: {response.content[0].text}")


@messages.command()
# cmd: python -m samples.messages video_url
@click.command()
def video_url():
    """
    影片內容區塊 (M3 only)
    - 最大50 MB (512 MB via Files API)
    """
    client = get_client()
    response = client.messages.create(
        model="MiniMax-M3",
        max_tokens=1024,
        messages=[
            {
                "role": "user",
                "content": [
                    {"type": "text", "text": "What happens in this video?"},
                    {
                        "type": "video",
                        "source": {
                            "type": "url",
                            "media_type": "video/mp4",
                            "url": "https://example.com/video.mp4",
                        },
                    },
                ],
            }
        ],
    )
    click.echo(f"Response: {response.content[0].text}")


@messages.command()
# cmd: python -m samples.messages video_mmfile
@click.command()
def video_mmfile():
    """使用 mm_file://引用已上傳的檔案"""
    client = get_client()
    response = client.messages.create(
        model="MiniMax-M3",
        max_tokens=1024,
        messages=[
            {
                "role": "user",
                "content": [
                    {"type": "text", "text": "What happens in this video?"},
                    {
                        "type": "video",
                        "source": {
                            "type": "mm_file",
                            "media_type": "video/mp4",
                            "data": "file_id_12345",
                        },
                    },
                ],
            }
        ],
    )
    click.echo(f"Response: {response.content[0].text}")


if __name__ == "__main__":
    messages()

File: cmd/sample/api/messages.go
package main

import (
	"context"
	"encoding/base64"
	"fmt"

	"github.com/anthropics/anthropic-sdk-go"
)

// Messages 示範各種訊息內容區塊的組裝方式。
// 對應 SDK 文件: Content blocks（文字、圖片 URL、圖片 base64）。
func Messages() error {
	if err := multiTextBlock(); err != nil {
		return fmt.Errorf("multiTextBlock: %w", err)
	}
	if err := imageURLBlock(); err != nil {
		return fmt.Errorf("imageURLBlock: %w", err)
	}
	if err := imageBase64Block(); err != nil {
		return fmt.Errorf("imageBase64Block: %w", err)
	}
	return nil
}

// multiTextBlock — 同一訊息內放入多個文字區塊
func multiTextBlock() error {
	fmt.Println("[messages.multiTextBlock]")
	client := getClient()
	message, err := client.Messages.New(context.TODO(), anthropic.MessageNewParams{
		Model:     defaultModel,
		MaxTokens: 1024,
		Messages: []anthropic.MessageParam{
			{
				Role: anthropic.MessageParamRoleUser,
				Content: []anthropic.ContentBlockParamUnion{
					{OfText: &anthropic.TextBlockParam{Text: "What is the capital of France?"}},
					{OfText: &anthropic.TextBlockParam{Text: "What about Germany?"}},
				},
			},
		},
	})
	if err != nil {
		return err
	}
	for _, b := range message.Content {
		if t, ok := b.AsAny().(anthropic.TextBlock); ok {
			fmt.Println("  ", t.Text)
		}
	}
	return nil
}

// imageURLBlock — 透過 URL 引用遠端圖片
// 支援格式: JPEG / PNG / GIF / WEBP；單檔上限 10 MB
func imageURLBlock() error {
	fmt.Println("\n[messages.imageURLBlock]")
	client := getClient()
	message, err := client.Messages.New(context.TODO(), anthropic.MessageNewParams{
		Model:     defaultModel,
		MaxTokens: 1024,
		Messages: []anthropic.MessageParam{
			{
				Role: anthropic.MessageParamRoleUser,
				Content: []anthropic.ContentBlockParamUnion{
					{OfText: &anthropic.TextBlockParam{Text: "What is in this image?"}},
					{OfImage: &anthropic.ImageBlockParam{
						Source: anthropic.ImageBlockParamSourceUnion{
							OfURL: &anthropic.URLImageSourceParam{
								URL: "https://upload.wikimedia.org/wikipedia/commons/thumb/0/0c/GoldenGateBridge-001.jpg/1200px-GoldenGateBridge-001.jpg",
							},
						},
					}},
				},
			},
		},
	})
	if err != nil {
		return err
	}
	for _, b := range message.Content {
		if t, ok := b.AsAny().(anthropic.TextBlock); ok {
			fmt.Println("  ", t.Text)
		}
	}
	return nil
}

// imageBase64Block — 直接以 base64 編碼內嵌小型圖片
// 範例使用 1x1 透明 PNG；實務上會用 os.ReadFile 讀入真實檔案
func imageBase64Block() error {
	fmt.Println("\n[messages.imageBase64Block]")
	client := getClient()
	// 1x1 透明 PNG
	const png1x1 = "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg=="
	data, _ := base64.StdEncoding.DecodeString(png1x1)

	message, err := client.Messages.New(context.TODO(), anthropic.MessageNewParams{
		Model:     defaultModel,
		MaxTokens: 1024,
		Messages: []anthropic.MessageParam{
			{
				Role: anthropic.MessageParamRoleUser,
				Content: []anthropic.ContentBlockParamUnion{
					{OfText: &anthropic.TextBlockParam{Text: "Describe this image."}},
					{OfImage: &anthropic.ImageBlockParam{
						Source: anthropic.ImageBlockParamSourceUnion{
							OfBase64: &anthropic.Base64ImageSourceParam{
								MediaType: anthropic.Base64ImageSourceMediaTypeImagePNG,
								Data:      string(data),
							},
						},
					}},
				},
			},
		},
	})
	if err != nil {
		return err
	}
	for _, b := range message.Content {
		if t, ok := b.AsAny().(anthropic.TextBlock); ok {
			fmt.Println("  ", t.Text)
		}
	}
	return nil
}
